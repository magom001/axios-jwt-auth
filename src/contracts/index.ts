export type Token = string;

export interface IAuthTokens {
  accessToken: Token;
  refreshToken: Token;
}

export interface ITokensStorage {
  saveTokens(tokens: IAuthTokens): Promise<void>;
  clearTokens(): Promise<void>;
  getAccessToken(): Promise<Token>;
  getRefreshToken(): Promise<Token>;
}