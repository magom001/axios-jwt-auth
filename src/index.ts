export * from './interfaces';
export { applyInterceptors } from './interceptors';
