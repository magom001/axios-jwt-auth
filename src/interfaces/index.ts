import { AxiosError, InternalAxiosRequestConfig } from 'axios';

export type Token = string;

export interface IAuthTokens {
  accessToken: Token;
  refreshToken: Token;
}

export interface ITokensStorage {
  saveTokens(tokens: IAuthTokens): Promise<void>;
  clearTokens(): Promise<void>;
  getAccessToken(): Promise<Token>;
  getRefreshToken(): Promise<Token>;
}

export interface Config {
  /**
   * Provide a custom function to attach access token to the outcoming request.
   * By default Bearer Authentication is used.
   *
   * @param {InternalAxiosRequestConfig} requestConfig
   */
  applyAccessToken?: (requestConfig: InternalAxiosRequestConfig, accessToken: Token) => void;
  /**
   * Provide a custom storage for the tokens. Should implement ITokensStorage interfaces.
   * Default to localStorage.
   */
  tokensStorage?: ITokensStorage;
  /**
   * A callback called to refresh tokens. The response should extend IAuthTokens interface.
   *
   * @param {Token} refreshToken
   */
  refreshTokens(refreshToken: Token): Promise<IAuthTokens>;
  /**
   * A callback that accepts an AxiosError and resolves into a boolean.
   * By default will refresh tokens on 401 status code.
   *
   * @param {AxiosError} error
   * @returns {Promise<boolean>}
   */
  shouldRefresh?: (error: AxiosError) => Promise<boolean>;

  /**
   * Optional callback fired when refresh requests fails.
   *
   * @param {AxiosError} error
   */
  onFailedToRefresh?: (error: AxiosError) => void;
}
