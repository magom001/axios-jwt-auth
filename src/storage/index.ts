import { IAuthTokens, ITokensStorage } from '../contracts';

class LocalStorageTokensStorage implements ITokensStorage {
  private static ACCESS_KEY_LS_KEY = 'accessToken';
  private static REFRESH_KEY_LS_KEY = 'refreshToken';

  async saveTokens(tokens: IAuthTokens): Promise<void> {
    localStorage.setItem(LocalStorageTokensStorage.ACCESS_KEY_LS_KEY, tokens.accessToken);
    localStorage.setItem(LocalStorageTokensStorage.REFRESH_KEY_LS_KEY, tokens.refreshToken);
  }

  async clearTokens(): Promise<void> {
    localStorage.removeItem(LocalStorageTokensStorage.ACCESS_KEY_LS_KEY);
    localStorage.removeItem(LocalStorageTokensStorage.REFRESH_KEY_LS_KEY);
  }

  async getAccessToken(): Promise<string> {
    return localStorage.getItem(LocalStorageTokensStorage.ACCESS_KEY_LS_KEY) ?? '';
  }

  async getRefreshToken(): Promise<string> {
    return localStorage.getItem(LocalStorageTokensStorage.REFRESH_KEY_LS_KEY) ?? '';
  }
}

export const defaultTokensStorage = new LocalStorageTokensStorage();
