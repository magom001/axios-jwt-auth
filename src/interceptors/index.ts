import { AxiosError, AxiosInstance, AxiosRequestConfig } from 'axios';

import { Token, Config } from '../contracts';
import { defaultTokensStorage } from '../storage';

let refreshPromise: Promise<Token> | null = null;

const defaultApplyAccessToken: Config['applyAccessToken'] = (requestConfig, token) => {
  if (requestConfig.headers) {
    requestConfig.headers.Authorization = `Bearer ${token}`;
  } else {
    requestConfig.headers = { Authorization: `Bearer ${token}` };
  }
};

const defaultShouldRefresh: Config['shouldRefresh'] = async (error: AxiosError) => {
  return error.response?.status === 401;
};

const requestInterceptor =
  ({ applyAccessToken = defaultApplyAccessToken, tokensStorage = defaultTokensStorage }: Config) =>
  async (requestConfig: AxiosRequestConfig): Promise<AxiosRequestConfig> => {
    let accessToken = await tokensStorage.getAccessToken();

    if (refreshPromise) {
      accessToken = await refreshPromise;
    }

    applyAccessToken(requestConfig, accessToken);

    return requestConfig;
  };

const responseErrorInterceptor =
  (
    axios: AxiosInstance,
    {
      tokensStorage = defaultTokensStorage,
      shouldRefresh = defaultShouldRefresh,
      refreshTokens,
      onFailedToRefresh,
    }: Config,
  ) =>
  async (error: AxiosError) => {
    if (await shouldRefresh(error)) {
      if (refreshPromise) {
        await refreshPromise;

        return axios.request(error.config);
      }

      refreshPromise = new Promise(async (resolve, reject) => {
        try {
          const refreshToken = await tokensStorage.getRefreshToken();
          if (!refreshToken) {
            throw error;
          }

          const tokens = await refreshTokens(refreshToken);

          await tokensStorage.saveTokens(tokens);

          resolve(tokens.accessToken);
        } catch (e) {
          await tokensStorage.clearTokens();

          reject(e);

          onFailedToRefresh?.(e as AxiosError<any>);
        } finally {
          refreshPromise = null;
        }
      });

      await refreshPromise;

      return axios.request(error.config);
    }

    throw error;
  };

/**
 * @param {AxiosInstance} axios
 * @param {Config} config
 */
export const applyInterceptors = (axios: AxiosInstance, config: Config): void => {
  axios.interceptors.request.use(requestInterceptor(config));
  axios.interceptors.response.use(undefined, responseErrorInterceptor(axios, config));
};
