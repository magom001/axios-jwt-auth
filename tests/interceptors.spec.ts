import axios from 'axios';
import nock from 'nock';

import { IAuthTokens, ITokensStorage, Token } from '../src/contracts';
import { applyInterceptors } from '../src/interceptors';

class InMemoryTokensStorage implements ITokensStorage {
  private accessToken = '';
  private refreshToken = '';

  async saveTokens(tokens: IAuthTokens): Promise<void> {
    this.accessToken = tokens.accessToken;
    this.refreshToken = tokens.refreshToken;
  }

  async clearTokens(): Promise<void> {
    this.accessToken = '';
    this.refreshToken = '';
  }

  async getAccessToken(): Promise<string> {
    return this.accessToken;
  }

  async getRefreshToken(): Promise<string> {
    return this.refreshToken;
  }
}

const baseURL = 'https://localhost:9999';

describe('JWT interceptor', () => {
  const getAxiosInstance = () => {
    const tokensStorage = new InMemoryTokensStorage();

    const refreshTokens = async (token: Token) => {
      const { data } = await axios.post<IAuthTokens>(`${baseURL}/api/refresh`, { token });

      return data;
    };

    const axiosInstance = axios.create({ baseURL });
    applyInterceptors(axiosInstance, { refreshTokens, tokensStorage });

    return { axiosInstance, tokensStorage };
  };

  it('should apply access token', async () => {
    const accessToken = 'xxxx-xxxx-xxxx-xxxx';
    const scope = nock(baseURL)
      .get('/api/test')
      .matchHeader('authorization', `Bearer ${accessToken}`)
      .reply(200, { ok: 'ok' });

    const { axiosInstance, tokensStorage } = getAxiosInstance();

    tokensStorage.saveTokens({ accessToken, refreshToken: '' });

    const { data } = await axiosInstance.get('/api/test');

    expect(data).toEqual({ ok: 'ok' });
    expect(scope.isDone()).toBe(true);
  });

  it('should refresh tokens when 401 status is received', async () => {
    const accessToken = 'xxxx-xxxx-xxxx-xxxx';
    const refreshToken = 'xxxx-refresh-xxxx';
    const newTokens = {
      accessToken: 'new-access-token',
      refreshToken: 'new-refresh-token',
    };

    const scope = nock(baseURL)
      .get('/api/test')
      .reply(401)
      .post('/api/refresh')
      .reply(200, newTokens)
      .get('/api/test')
      .matchHeader('authorization', `Bearer ${newTokens.accessToken}`)
      .reply(200, { ok: 'ok' });

    const { axiosInstance, tokensStorage } = getAxiosInstance();

    tokensStorage.saveTokens({ accessToken, refreshToken });

    const { data } = await axiosInstance.get('/api/test');

    expect(data).toEqual({ ok: 'ok' });
    expect(scope.isDone()).toBe(true);
    expect(await tokensStorage.getRefreshToken()).toEqual(newTokens.refreshToken);
    expect(await tokensStorage.getAccessToken()).toEqual(newTokens.accessToken);
  });

  it('new requests should wait till refresh is resolved', async () => {
    const accessToken = 'xxxx-xxxx-xxxx-xxxx';
    const refreshToken = 'xxxx-refresh-xxxx';
    const newTokens = {
      accessToken: 'new-access-token',
      refreshToken: 'new-refresh-token',
    };
    const scope = nock(baseURL)
      .get('/api/test')
      .reply(401)
      .post('/api/refresh')
      .delay(300)
      .reply(200, newTokens)
      .get('/api/test')
      .matchHeader('authorization', `Bearer ${newTokens.accessToken}`)
      .thrice()
      .reply(200, { ok: 'ok' });

    const { axiosInstance, tokensStorage } = getAxiosInstance();

    tokensStorage.saveTokens({ accessToken, refreshToken });

    axiosInstance.get('/api/test');

    await new Promise((r) => setTimeout(r, 100));

    await Promise.all([axiosInstance.get('/api/test'), axiosInstance.get('/api/test')]);

    expect(scope.isDone()).toBe(true);
  });

  it('should refresh only once when several parallel requests resolve with 401', async () => {
    const accessToken = 'xxxx-xxxx-xxxx-xxxx';
    const refreshToken = 'xxxx-refresh-xxxx';
    const newTokens = {
      accessToken: 'new-access-token',
      refreshToken: 'new-refresh-token',
    };
    const scope = nock(baseURL)
      .get('/api/test')
      .thrice()
      .reply(401)
      .post('/api/refresh')
      .reply(200, newTokens)
      .get('/api/test')
      .matchHeader('authorization', `Bearer ${newTokens.accessToken}`)
      .thrice()
      .reply(200, { ok: 'ok' });

    const { axiosInstance, tokensStorage } = getAxiosInstance();

    tokensStorage.saveTokens({ accessToken, refreshToken });

    await Promise.all([
      axiosInstance.get('/api/test'),
      axiosInstance.get('/api/test'),
      axiosInstance.get('/api/test'),
    ]);

    expect(scope.isDone()).toBe(true);
  });

  it('should reject requests if refresh fails', async () => {
    expect.assertions(4);

    const accessToken = 'xxxx-xxxx-xxxx-xxxx';
    const refreshToken = 'xxxx-refresh-xxxx';

    const scope = nock(baseURL)
      .get('/api/test')
      .thrice()
      .reply(401)
      .post('/api/refresh')
      .reply(500);

    const { axiosInstance, tokensStorage } = getAxiosInstance();

    tokensStorage.saveTokens({ accessToken, refreshToken });

    const requests = [
      axiosInstance.get('/api/test'),
      axiosInstance.get('/api/test'),
      axiosInstance.get('/api/test'),
    ];

    for (const r of requests) {
      await expect(r).rejects.toMatchSnapshot();
    }

    expect(scope.isDone()).toBe(true);
  });

  it('should reject queued requests if refresh fails', async () => {
    expect.assertions(4);

    const accessToken = 'xxxx-xxxx-xxxx-xxxx';
    const refreshToken = 'xxxx-refresh-xxxx';

    const scope = nock(baseURL)
      .get('/api/test')
      .reply(401)
      .post('/api/refresh')
      .delay(300)
      .reply(500);

    const { axiosInstance, tokensStorage } = getAxiosInstance();

    tokensStorage.saveTokens({ accessToken, refreshToken });

    const initial = axiosInstance.get('/api/test');

    await new Promise((r) => setTimeout(r, 100));

    const requests = [initial, axiosInstance.get('/api/test'), axiosInstance.get('/api/test')];

    for (const r of requests) {
      await expect(r).rejects.toMatchSnapshot();
    }

    expect(scope.isDone()).toBe(true);
  });
});
