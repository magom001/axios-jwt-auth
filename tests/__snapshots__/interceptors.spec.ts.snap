// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JWT interceptor should reject queued requests if refresh fails 1`] = `[AxiosError: Request failed with status code 500]`;

exports[`JWT interceptor should reject queued requests if refresh fails 2`] = `[AxiosError: Request failed with status code 500]`;

exports[`JWT interceptor should reject queued requests if refresh fails 3`] = `[AxiosError: Request failed with status code 500]`;

exports[`JWT interceptor should reject requests if refresh fails 1`] = `[AxiosError: Request failed with status code 500]`;

exports[`JWT interceptor should reject requests if refresh fails 2`] = `[AxiosError: Request failed with status code 500]`;

exports[`JWT interceptor should reject requests if refresh fails 3`] = `[AxiosError: Request failed with status code 500]`;
